python-sdk/nuscenes/predict/models/mtp.py

```diff
@@ -0,0 +1,328 @@
+        return backbone_features.mean([2, 3])
+
+
+class MTP(nn.Module):
+    """ Implements the MTP network. """
+
+    def __init__(self, backbone: nn.Module, num_modes: int,
+                 seconds: float = 6, frequency_in_hz: float = 2,
+                 n_hidden_layers: int = 4096, input_shape: Tuple[int, int, int] = (3, 500, 500)):
+        """
+        Inits the MTP network.
+        :param backbone: CNN Backbone to use.
+        :param num_modes: Number of predicted paths to estimate for each agent.
+        :param seconds: Number of seconds into the future to predict.
+            Default for the challenge is 6.
+        :param frequency_in_hz: Frequency between timesteps in the prediction (in Hz).
+            Highest frequency is nuScenes is 2 Hz.
+        :param n_hidden_layers: Size of fully connected layer after the CNN
+            backbone processes the image.
+        :param input_shape: Shape of the input expected by the network.
+            This is needed because the size of the fully connected layer after
+            the backbone depends on the backbone and its version.
+
+        Note:
+            Although seconds and frequency_in_hz are labeled their
+            product should be an int.
+        """
+
+        super().__init__()
+
+        self.backbone = backbone
+        self.num_modes = num_modes
+        backbone_feature_dim = self._calculate_backbone_feature_dim(input_shape)
+        self.fc1 = nn.Linear(backbone_feature_dim + 3, n_hidden_layers)
+        predictions_per_mode = int(seconds * frequency_in_hz) * 2
+
+        self.fc2 = nn.Linear(n_hidden_layers, int(num_modes * predictions_per_mode + num_modes))
+
+    def _calculate_backbone_feature_dim(self, input_shape: Tuple[int, int, int]) -> int:
+        """ Helper to calculate the shape of the fully-connected regression layer. """
+        tensor = torch.ones(1, *input_shape)
+        output_feat = self.backbone.forward(tensor)
+        return output_feat.shape[-1]
+
+    def forward(self, image_tensor: torch.Tensor,
+                agent_state_vector: torch.Tensor) -> torch.Tensor:
+        """
+        Forward pass of the model.
+        """
+
+        backbone_features = self.backbone(image_tensor)
+
+        features = torch.cat([backbone_features, agent_state_vector], axis=1)
+
+        predictions = self.fc2(self.fc1(features))
+
+        # Normalize the probabilities to sum to 1 for inference.
+        mode_probabilities = predictions[:, -self.num_modes:].clone()
+        if not self.training:
+            mode_probabilities = f.softmax(mode_probabilities, dim=-1)
+
+        predictions = predictions[:, :-self.num_modes]
+
+        return torch.cat((predictions, mode_probabilities), 1)
+
+
+class MTPLoss:
+    """
+    Computes the loss for the MTP model.
+    """
+
+    def __init__(self,
+                 num_modes: int,
+                 regression_loss_weight: float,
+                 angle_threshold_degrees: float = 5.):
```

Comment:
@freddyaboulton We've been using `7` as the default value for the `angle_threshold_degrees` param for our models.

Reply:
Thanks! Think I'll keep it at five since that's what they mentioned in the paper.